src/windows/minidump_writer.rs

```diff
@@ -0,0 +1,279 @@
+use crate::windows::errors::Error;
+use minidump_common::format::{BreakpadInfoValid, MINIDUMP_BREAKPAD_INFO, MINIDUMP_STREAM_TYPE};
+use scroll::Pwrite;
+use std::{ffi::c_void, os::windows::io::AsRawHandle};
+pub use windows_sys::Win32::Foundation::HANDLE;
+use windows_sys::Win32::{
+    Foundation::{CloseHandle, ERROR_SUCCESS, STATUS_INVALID_HANDLE},
+    System::{
+        ApplicationVerifier as av,
+        Diagnostics::Debug as md,
+        Threading::{GetCurrentThreadId, OpenProcess},
+    },
+};
+
+// For some reason this is defined in SystemServices which is massive and not worth bringing in for ONE constant
+const GENERIC_ALL: u32 = 268435456;
+
+pub struct MinidumpWriter {
+    /// The crash context as captured by an exception handler
+    crash_context: crash_context::CrashContext,
+    /// Handle to the crashing process, which could be ourselves
+    crashing_process: HANDLE,
+    /// The pid of the crashing process.
+    crashing_pid: u32,
+    /// The `EXCEPTION_POINTERS` contained in crash context is a pointer into the
+    /// memory of the process that crashed, as it contains an `EXCEPTION_RECORD`
+    /// record which is an internally linked list, so in the case that we are
+    /// dumping a process other than the current one, we need to tell
+    /// MiniDumpWriteDump that the pointers come from an external process so that
+    /// it can use eg ReadProcessMemory to get the contextual information from
+    /// the crash, rather than from the current process
+    is_external_process: bool,
+}
+
+impl MinidumpWriter {
+    /// Creates a minidump writer for a crash that occurred in an external process.
+    pub fn external_process(
+        crash_context: crash_context::CrashContext,
+        pid: u32,
+    ) -> Result<Self, Error> {
+        let crashing_process = unsafe {
+            OpenProcess(
+                GENERIC_ALL, // desired access
+                0,           // inherit handles
+                pid,         // pid
+            )
+        };
+
+        if crashing_process == 0 {
+            Err(std::io::Error::last_os_error().into())
+        } else {
+            Ok(Self {
+                crash_context,
+                crashing_process,
+                crashing_pid: pid,
+                is_external_process: true,
+            })
+        }
+    }
+
+    /// Creates a minidump writer for a crash that occurred in the current process.
+    ///
+    /// # Errors
+    ///
+    /// Fails if we are unable to open a `HANDLE` to the current process
+    pub fn current_process(crash_context: crash_context::CrashContext) -> Result<Self, Error> {
+        let crashing_pid = std::process::id();
+
+        // SAFETY: syscall
+        let crashing_process = unsafe {
```

Comment:
You could use [`GetCurrentProcess`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentprocess) instead. It never fails and always has `PROCESS_ALL_ACCESS`